OrbitSsh/Channel.cpp

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitSsh/Channel.h"
+
+#include <libssh2.h>
+
+#include <memory>
+#include <optional>
+#include <outcome.hpp>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitSsh/ResultType.h"
+
+namespace OrbitSsh {
+
+Channel::Channel(Session* session_ptr)
+    : session_ptr_(session_ptr), raw_channel_ptr_(nullptr) {}
+
+Channel::~Channel() { libssh2_channel_free(raw_channel_ptr_); }
+
+ResultType Channel::Read(std::string* result) {
+  int rc;
+  bool read_at_least_once = false;
+  int buffer_size = sizeof(buffer_);
+  do {
+    rc = libssh2_channel_read(raw_channel_ptr_, buffer_, buffer_size);
+
+    if (rc == LIBSSH2_ERROR_EAGAIN) break;
+    if (rc < 0) {
+      ERROR("Unable to read channel, error code %d", rc);
+      return kError;
+    }
+
+    *result += std::string(buffer_, rc);
+    read_at_least_once = true;
+  } while (rc == buffer_size);
+
+  if (read_at_least_once) return kSuccess;
+  return kAgain;
+}
+
+ResultType Channel::Write(const std::string& text, int* written_length) {
```

Comment:
That API is a little bit weird. Why not return the written_length? For the consumer it's also unclear that written_length is also an IN-parameter.

Reply:
This function is supposed to be private. -> fixed